### ISSUE TITLE
Consider deleting the CoreMangLib tests and instead running a binary drop of corresponding CoreFX tests

### ISSUE THREAD

**Issue by tannergooding** [dotnet/coreclr#12782]
As per the discussion here: https://github.com/dotnet/coreclr/pull/12755#issuecomment-314896715.

The CoreMangLib represent a set of tests which often duplicate test coverage that exists in CoreFX. Additionally, the subset of tests that are duplicated represent a very small portion of the actual code-base.

We should consider deleting the CoreMangLib tests and running a binary drop of CoreFX tests that validate corlib functionality for PRs.

**Comment by RussKeldorph** [comment]
- [ ] Move the baseline CoreFx jobs ([example](https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x64_checked_windows_nt_corefx_baseline/)) to root
 - [ ] Make one or CoreFX job PR-triggered
 - [ ] Delete CoreMangLib tests
 - [ ] Profit

**Comment by jkotas** [comment]
The current CoreFx jobs are fragile. They fail when  CoreFX gets out of sync - it relatively often by design, or when regression sneaks into CoreFX. I do not think we want to be building upon fragile solutions like that.

I think what needs to happen before the above steps is:
[ ] CoreFX official builds should produce package with CoreFX test binaries
[ ] CoreCLR repo consumes this test package via https://github.com/dotnet/coreclr/blob/master/dependencies.props
[ ] baseline CoreFx jobs are switched to use this test package

**Comment by RussKeldorph** [comment]
@jkotas Fair enough, though arguably the fragility should be addressed regardless of this issue.  My understanding is that your solution doesn't address some of the "by design" breaks, though.  I believe we'd actually need a staging branch of CoreFx where breaking changes can be made, and have CoreCLR's tests depend on that rather than master.  Producing packages for that branch may be more complex than referring to a git branch, even though the former does have obvious other advantages.

**Comment by jkotas** [comment]
The per-PR Windows legs do not have this fragility today. And we are fixing this fragility for the per-PR Unix legs. My point was that it would be pretty unfortunate to introduce this fragility again across both Windows and Unix.

I do not think we would need staging branch. We do not have staging branch for CoreFX product binaries either. The CoreFX test package would be always the same version as the CoreFX product binaries package that CoreCLR depends on.

To deal with breaking changes, we may just need a file to disable selected CoreFX tests temporarily until the breaking change propagates through the system.

**Comment by RussKeldorph** [comment]
Yes, I understand there is a clear dependency of this issue on stabilizing the corefx jobs in order to change the PR triggers.  Sorry that wasn't clear and thank you for pointing that out.  I just meant to say that stabilizing corefx jobs ought to be its own issue for other reasons.

I think there may be multiple features we need to add to the CoreFx test harness in addition to packaging them.  For example, last time I checked, they didn't have a supported (or at least obvious) way to run all the tests without building them first.

**Comment by RussKeldorph** [comment]
Also, if we don't need the staging branch, then I believe we could achieve the proposed stability today by simply making the current CoreFX jobs discover the commit corresponding to the CoreFX product packages and use that instead of tip.  Inability to selectively disable tests and time wasted building tests would need to be addressed, though.

I'm not saying we shouldn't package the CoreFX tests; I just want to note the possible options.

**Comment by jkotas** [comment]
Agree that there are options. I would love to be in a place where we have consistent solution for both CoreFX product binaries and CoreFX tests (if we use package for one, we should use package for the other; if we use build from source for one, we should use build from source for the other).

**Comment by danmosemsft** [comment]
Now https://github.com/dotnet/coreclr/pull/15944 is submitted (cc @BruceForstall @RussKeldorph ), can we now proceed to delete CoreMangLib tests? 

How much attention needs to be paid to them to make sure there aren't a few jewels in the dross eg a test to cover some particular bug we fixed, that is not represented in the CoreFX test bed? Eg I saw https://github.com/dotnet/coreclr/blob/master/tests/src/CoreMangLib/system/delegate/regressions/Github_13160/Github_13160.cs. It does seem the vast majority is junk though.

**Comment by danmosemsft** [comment]
Hmm, I see from the discussion that after #15944 we still have cases where tests will not always 100% pass. Do we believe it's important to keep CoreMangLib tests meantime - I assume we have to?

**Comment by jkotas** [comment]
And I am also worried about the CI cost when everything starts downloading and building CoreFX.

**Comment by BruceForstall** [comment]
@danmosemsft Also, while #15944 solves some problems, @jkotas has convinced me that it's not really the right or full solution.

**Comment by danmosemsft** [comment]
Is the primary purpose of the CoreMangLib tests to protect corelib, or to help exercise changes elsewhere (VM, GC, JIT)?

Either way perhaps what is needed is a selected subset of tests forked from CoreFX (or mirroring might work since this subset will less often have breaking changes). 

But, if CoreMangLib performs this job well enough for now and is not expensive to maintain, maybe it is not worth bothering with.

**Comment by jkotas** [comment]
> Is the primary purpose of the CoreMangLib tests to protect corelib, or to help exercise changes elsewhere 

Both.

I think that the primary problem with CoreMangLib is that it is expensive to build and run. Poor cost / value.

**Comment by BruceForstall** [comment]
We still use the "old mechanism" for running corefx jobs for all non-PR triggered corefx test jobs. Those clone the live corefx repo, build it, and run tests. We have that (now, as of a few days ago) for all architectures, and many JIT stress modes.

Thus, if the CoreMangLib directory is a subset of the full corefx repo, I don't think we'll lose any coverage because we have our existing pri-1 corefx jobs.

**Comment by danmosemsft** [comment]
Yay! Thanks @jkotas!
